### PR TITLE
[BEAM-646][BEAM-1173] Remove PipelineRunner#apply

### PR DESCRIPTION
**Pull request by tgroh**
All existing Pipeline Runners that use the Java SDK modify Pipeline
graphs with the Pipeline Surgery APIs. Apply is now superflous.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles 
CC: @amitsela @tweise @aljoscha

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AssertionCountingVisitor.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  private AssertionCountingVisitor() {
+    assertCount = 0;
+    pipelineVisited = false;
+  }
+
+  @Override
+  public CompositeBehavior enterCompositeTransform(Node node) {
+    if (node.isRootNode()) {
+      checkState(
```
Comment (kennknowles) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AssertionCountingVisitor.java]: Given this restriction, how about making it private and exposing the result of the traversal via a method like `PAssert.countAssertsInPipeline(Pipeline)`. It doesn't _quite_ belong in the core SDK, but as long as it is private it seems fine. Would that even allow the various `PAssert` transforms to become private?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 2]
```diff
@@ -88,37 +94,24 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
...
+      adaptBoundedReads(pipeline);
     }
-  }
-
-  @Override
-  public SparkPipelineResult run(Pipeline pipeline) {
-    TestSparkPipelineOptions testSparkPipelineOptions =
-        pipeline.getOptions().as(TestSparkPipelineOptions.class);
     SparkPipelineResult result = null;
     // clear state of Aggregators, Metrics and Watermarks.
```
Comment (tgroh) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: Fix't.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 3]
```diff
@@ -88,37 +94,24 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
...
+    AssertionCountingVisitor assertionCountingVisitor = AssertionCountingVisitor.create();
```
Comment (tgroh) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: Neato. Done.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 4]
```diff
@@ -88,37 +94,24 @@ public static TestSparkRunner fromOptions(PipelineOptions options) {
...
+    pipeline.traverseTopologically(assertionCountingVisitor);
+    expectedNumberOfAssertions = assertionCountingVisitor.getPAssertCount();
```
Comment (tgroh) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java]: Done.

File: runners/spark/pom.xml [hunk 5]
```diff
@@ -325,6 +336,10 @@
       <type>test-jar</type>
       <scope>test</scope>
     </dependency>
+      <dependency>
```
Comment (tgroh) [inline runners/spark/pom.xml]: rm'd this one

File: runners/spark/pom.xml [hunk 6]
```diff
@@ -226,6 +226,17 @@
     </dependency>
     <dependency>
       <groupId>org.apache.beam</groupId>
+      <artifactId>beam-runners-core-construction-java</artifactId>
```
Comment (tgroh) [inline runners/spark/pom.xml]: rm'd that exclusion

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AssertionCountingVisitor.java (hunk 1 above)
Comment (tgroh) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/AssertionCountingVisitor.java]: I prefer the former factoring, certainly. Done.

For the latter, I'll check; but is definitely a follow-up PR if possible.

**Comment by kennknowles** [comment]
LGTM! (remember to squash)

**Comment by amitsela** [comment]
LGTM

**Pull request closed without merging**